File: src/karatsubaMultiplication/karatsubaMultiplication.java
package karatsubaMultiplication;

import static java.lang.Math.*;

public class karatsubaMultiplication {
    public static long multiplyInts(long int1, long int2) {
        if (int1 < 10 || int2 < 10) {
            return int1 * int2;
        }

        String str1 = Long.toString(int1);
        String str2 = Long.toString(int2);
        int len1 = str1.length();
        int len2 = str2.length();

        int power = (int) floor((float) min(len1, len2) / 2);

        long a = Long.parseLong(str1.substring(0, len1-power));
        long b = Long.parseLong(str1.substring(len1-power));
        long c = Long.parseLong(str2.substring(0, len2-power));
        long d = Long.parseLong(str2.substring(len2-power));

        long ac = multiplyInts(a, c);
        long ad_plus_bc = multiplyInts(a+b, c+d);
        long bd = multiplyInts(b, d);

        return (long) (pow(10, power * 2) * ac + pow(10, power) * (ad_plus_bc - ac - bd) + bd);
    }

    public static void main(String[] args) {
        long int1 = 31415926L;
        long int2 = 27182818L;
        long result = multiplyInts(int1, int2);
        long built_in = int1 * int2;
        System.out.println("Using Karatsuba: " + result);
        System.out.println("Using built-in: " + built_in);
        if (result == built_in) {
            System.out.println("Results match");
        }
        else {
            System.out.println("Results do not match");
        }
    }
}

File: Karatsuba_Multiplication/karatsuba_multiplication.py
import math


def multiply_ints(int1, int2):
    if int1 < 10 or int2 < 10:
        return int1 * int2

    str1 = str(int1)
    str2 = str(int2)
    len1 = len(str1)
    len2 = len(str2)

    power = math.floor(min(len1, len2) / 2)

    a = int(str1[:len1-power])
    b = int(str1[len1-power:])
    c = int(str2[:len2-power])
    d = int(str2[len2-power:])

    ac = multiply_ints(a, c)
    ad_plus_bc = multiply_ints(a+b, c+d)
    bd = multiply_ints(b, d)

    return (10 ** (power * 2)) * ac + (10 ** power) * (ad_plus_bc - ac - bd) + bd

int1 = 3141592653589793238462643383279502884197169399375105820974944592
int2 = 2718281828459045235360287471352662497757247093699959574966967627

result = multiply_ints(int1, int2)
built_in = int1 * int2
print('Using Karatsuba: ', result)
print('Using built-in: ', built_in)
if result == built_in:
    print('Results match')
else:
    print('Results do not match')
